### ISSUE TITLE
Improve documentation

### ISSUE THREAD

**Issue by bernd-wechner** [dimagi/django-cte#44]
I am so pleased to found this, as it helped me solve a hairy problem that Django was stuck on (Windowing on a Windowed field). I've long wanted in Django a feature that can turn a QuerySet into the subject of another query.

But I admit I would like to understand what is going on here. I dump the SQL and like what I see, and it works for me, but I'm following the README scrip  without comprehension.

```py
cte = With(
    Order.objects
    .values("region_id")
    .annotate(total=Sum("amount"))
)

orders = (
    cte.join(Order, region=cte.col.region_id)
    .with_cte(cte)
    .annotate(region_total=cte.col.total)
    .order_by("amount")
)
```

What role do `With`, `.join` and `.with_cte` play here. It seems clunky syntax and there is no explanation of these component parts are doing.

As an ignoramus who has no clue what they do, I experimented of course, not least with the simple syntax I'd love to enjoy:

```py
orders = cte.annotate(region_total=cte.col.total).order_by("amount")
```

which fails, of course. 

I can see what the `.join()` does I guess in the resulting SQL, arises because the CTE is flexible and can be joined with other tables.

But could I for example select FROM the CTE? Without a join? 

The `.with_cte` puzzles me as cte is at the head and self to the join which could also provide same downstream, and it's a puzzle why `cte` is passed back into a cte chain. It would be nice to understand why.

On the whole, very happy to have found this, and it's been empowering, but I would like understanding as well if I can get to it. I may drill down into the code a little to get that.

**Comment by millerdev** [comment]
- `With(...)` constructs a common table expression to be placed in a SQL `WITH` clause.
- `.join(...)` adds a `JOIN` expression to the SQL `FROM` clause, joining the CTE query to some other queryset or model.
- `.with_cte(...)` adds a CTE constructed with `With(...)` to a `WITH` clause on the queryset.

Anatomy of the resulting SQL, slightly simplified for purposes of explanation.
```sql
WITH RECURSIVE "cte" AS (  -- WITH clause added by .with_cte(...)

    -- CTE query constructed in With(...)
    SELECT
        "tests_order"."region_id",
        SUM("tests_order"."amount") AS "total"
    FROM "tests_order"
        INNER JOIN "tests_region" ON (
            "tests_order"."region_id" = "tests_region"."name"
        )
    WHERE "tests_region"."parent_id" = 'sun'
    GROUP BY "tests_order"."region_id"
)
SELECT
    "tests_order"."id",
    "tests_order"."region_id",
    "tests_order"."amount",
    "cte"."total" AS "region_total"
FROM "tests_order"
INNER JOIN "cte" ON "tests_order"."region_id" = ("cte"."region_id")  -- JOIN clause added by .join(...)
ORDER BY "tests_order"."amount" ASC
```

The above query (except for comments and basic formatting) was produced by adding `print(orders.query)` to `TestCTE.test_simple_cte_query` and then running `nosetests tests/test_cte.py:TestCTE.test_simple_cte_query --nocapture`

Does that make things more clear?

**Comment by bernd-wechner** [comment]
Thanks, I worked that all out since and field a PR to simplify it a little by removing the need for `.with_cte().`. And I'll PR a README improvement some time too if you're content with the default CTE PR.

Another minor hassle I found the hard way along the way using this (otherwise brilliant addition to Django landscape - so brilliant it should be absorbed into Django as it solves a huge problem with Django namely querying on an encapsulated queryset) is that the result of the .join() includes only the fields from its argument and not the fields from the CTE, which need explicitly to be added again with an annotation as and when needed. That took me a long time to work out, and SQL dumps.

I might try and tidy that up a bit too. We shall see.
